File: camerax_platform_interface/lib/src/video/file_output_options.dart
import 'dart:io';

import 'package:camerax_platform_interface/src/common.dart';
import 'package:camerax_platform_interface/src/video.dart';
import 'package:plugin_platform_interface/plugin_platform_interface.dart';

abstract interface class FileOutputOptions$Builder {
  factory FileOutputOptions$Builder(File file) =>
      FileOutputOptionsChannel.instance.createBuilder(file);

  Future<FileOutputOptions$Builder> setDurationLimitMillis(
    int durationLimitMillis,
  );
  Future<FileOutputOptions$Builder> setFileSizeLimit(int fileSizeLimitBytes);
  Future<FileOutputOptions$Builder> setLocation(Location? location);
  Future<FileOutputOptions> build();
}

abstract interface class FileOutputOptions implements OutputOptions {
  Future<File> getFile();
}

abstract base class FileOutputOptionsChannel extends PlatformInterface {
  FileOutputOptionsChannel() : super(token: _token);

  static final _token = Object();

  static FileOutputOptionsChannel? _instance;

  static FileOutputOptionsChannel get instance =>
      ArgumentError.checkNotNull(_instance, 'instance');

  static set instance(FileOutputOptionsChannel instance) {
    PlatformInterface.verify(instance, _token);
    _instance = instance;
  }

  FileOutputOptions$Builder createBuilder(File file);
}

File: camerax_android/android/src/main/kotlin/dev/hebei/camerax_android/legacy/video/FileOutputOptions.kt
package dev.hebei.camerax_android.legacy.video

import dev.hebei.camerax_android.legacy.CameraXRegistrar
import dev.hebei.camerax_android.legacy.PigeonApiFileOutputOptions
import dev.hebei.camerax_android.legacy.common.args
import dev.hebei.camerax_android.legacy.common.fileObj

class FileOutputOptions(registrar: CameraXRegistrar) : PigeonApiFileOutputOptions(registrar) {
    override fun build(
        file: String, durationLimitMillis: Long?, fileSizeLimitBytes: Long?, location: android.location.Location?
    ): androidx.camera.video.FileOutputOptions {
        val builder = androidx.camera.video.FileOutputOptions.Builder(file.fileObj).setLocation(location)
        if (durationLimitMillis != null) {
            builder.setDurationLimitMillis(durationLimitMillis)
        }
        if (fileSizeLimitBytes != null) {
            builder.setFileSizeLimit(fileSizeLimitBytes)
        }
        return builder.build()
    }

    override fun file(pigeon_instance: androidx.camera.video.FileOutputOptions): String {
        return pigeon_instance.file.args
    }
}